From c3359886af6a8a788672d4d7b09973e52834a317 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:27:59 -0400
Subject: [alsa] hda: Fix loud click noise with IdeaPad 410Y

Message-id: <140972823573-233-patchset-jkysela@redhat.com>
Patchwork-id: 90933
O-Subject: [RHEL 7.1 PATCH 232/442] ALSA: hda - Fix loud click noise with IdeaPad 410Y
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

Lenovo IdeaPad 410Y with ALC282 codec makes loud click noises at boot
and shutdown.  Also, it wrongly misdetects the acpi_thinkpad hook.
This patch adds a device-specific fixup for disabling the shutup
callback that is the cause of the click noise and also avoiding the
thinpad_helper calls.

Bugzilla: https://bugzilla.kernel.org/show_bug.cgi?id=71511
Reported-and-tested-by: Guilherme Amadio <guilherme.amadio@gmail.com>
Cc: <stable@vger.kernel.org>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 9b745ab897199c2af9f21ca9681ef86d5b971002)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 86c01f0..6042971 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -3635,6 +3635,19 @@ static void alc_fixup_auto_mute_via_amp(struct hda_codec *codec,
  }
 }
 
+static void alc_no_shutup(struct hda_codec *codec)
+{
+}
+
+static void alc_fixup_no_shutup(struct hda_codec *codec,
+    const struct hda_fixup *fix, int action)
+{
+ if (action == HDA_FIXUP_ACT_PRE_PROBE) {
+  struct alc_spec *spec = codec->spec;
+  spec->shutup = alc_no_shutup;
+ }
+}
+
 static void alc_fixup_headset_mode_alc668(struct hda_codec *codec,
     const struct hda_fixup *fix, int action)
 {
@@ -3863,6 +3876,7 @@ enum {
  ALC269_FIXUP_HP_GPIO_LED,
  ALC269_FIXUP_INV_DMIC,
  ALC269_FIXUP_LENOVO_DOCK,
+ ALC269_FIXUP_NO_SHUTUP,
  ALC286_FIXUP_SONY_MIC_NO_PRESENCE,
  ALC269_FIXUP_PINCFG_NO_HP_TO_LINEOUT,
  ALC269_FIXUP_DELL1_MIC_NO_PRESENCE,
@@ -4041,6 +4055,10 @@ static const struct hda_fixup alc269_fixups[] = {
   .type = HDA_FIXUP_FUNC,
   .v.func = alc_fixup_inv_dmic_0x12,
  },
+ [ALC269_FIXUP_NO_SHUTUP] = {
+  .type = HDA_FIXUP_FUNC,
+  .v.func = alc_fixup_no_shutup,
+ },
  [ALC269_FIXUP_LENOVO_DOCK] = {
   .type = HDA_FIXUP_PINS,
   .v.pins = (const struct hda_pintbl[]) {
@@ -4440,6 +4458,7 @@ static const struct snd_pci_quirk alc269_fixup_tbl[] = {
  SND_PCI_QUIRK(0x17aa, 0x2212, "Thinkpad", ALC269_FIXUP_LIMIT_INT_MIC_BOOST),
  SND_PCI_QUIRK(0x17aa, 0x2214, "Thinkpad", ALC269_FIXUP_LIMIT_INT_MIC_BOOST),
  SND_PCI_QUIRK(0x17aa, 0x2215, "Thinkpad", ALC269_FIXUP_LIMIT_INT_MIC_BOOST),
+ SND_PCI_QUIRK(0x17aa, 0x3978, "IdeaPad Y410P", ALC269_FIXUP_NO_SHUTUP),
  SND_PCI_QUIRK(0x17aa, 0x5013, "Thinkpad", ALC269_FIXUP_LIMIT_INT_MIC_BOOST),
  SND_PCI_QUIRK(0x17aa, 0x501a, "Thinkpad", ALC283_FIXUP_INT_MIC),
  SND_PCI_QUIRK(0x17aa, 0x5026, "Thinkpad", ALC269_FIXUP_LIMIT_INT_MIC_BOOST),
-- 
1.7.1